crates/cli/src/server.rs

```diff
@@ -0,0 +1,208 @@
+    });
+
+    //return served file url.
+    let ret = files_data
+        .into_iter()
+        .map(|(path, (url, _hash))| (path, url))
+        .collect();
+    Ok((ret, jh))
+}
+
+async fn server_process_file(
+    req: Request<body::Incoming>,
+    files: &HashMap<String, PathBuf>,
+    counter_sender: tokio::sync::mpsc::UnboundedSender<(String, usize)>,
+) -> std::result::Result<Response<BoxBody<Bytes, std::io::Error>>, hyper::Error> {
+    let file_digest = &req.uri().path()[1..];
+    let file_path = match files.get(file_digest) {
+        Some(file_path) => file_path,
+        None => {
+            return Ok(Response::builder()
+                .status(StatusCode::NOT_FOUND)
+                .body(
+                    Full::new("Not found.".into())
+                        .map_err(|e| match e {})
+                        .boxed(),
+                )
+                .unwrap())
+        }
+    };
+    let file = match File::open(file_path).await {
+        Ok(file) => file,
+        Err(_) => {
+            return Ok(Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(
+                    Full::new("Internal server error".into())
+                        .map_err(|e| match e {})
+                        .boxed(),
+                )
+                .unwrap())
+        }
+    };
+
+    let file_hash = file_digest.to_string();
+    let reader = ReaderStream::new(file).map(move |bytes| {
+        let nb_bytes = bytes.as_ref().map(|b| b.len()).unwrap_or(0);
+        if let Err(err) = counter_sender.send((file_hash.clone(), nb_bytes)) {
+            log::error!("An error occurs during file download. NOtofication channel close:{err}");
+        }
+        bytes
+    });
+    let stream_body = StreamBody::new(reader.map_ok(Frame::data));
+
+    Ok(Response::builder()
+        .status(StatusCode::OK)
+        .body(BodyExt::boxed(stream_body))
+        .unwrap())
+}
+
+//calculate the file name digest and the local server url.
+//return the file path, file_url and file digest
+fn calculate_file_url_digest(file_path: &Path, listen_addr: SocketAddr) -> (String, String) {
+    //use to_string_lossy because path verification should have been done before.
+    let file_name = file_path.to_string_lossy();
+    let mut hasher = Sha3_256::new();
+    hasher.update(file_name.as_bytes());
+    let digest = hex::encode(hasher.finalize());
+    let file_url = format!("http://{}/{}", listen_addr, digest);
+    (file_url, digest)
+}
+
+fn build_file_progress_bar(file_name: String, total_size: u64) -> ProgressBar {
+    let pb = ProgressBar::new(total_size);
+    pb.set_style(ProgressStyle::with_template("{spinner:.green} [{elapsed_precise}] [{wide_bar:.cyan/blue}] {bytes}/{total_bytes} ({eta})")
+        .unwrap()
+        .with_key("eta", move |state: &ProgressState, w: &mut dyn Write| write!(w, "{}-{:.1}s", file_name, state.eta().as_secs_f64()).unwrap())
+        .progress_chars("#>-"));
+    pb
+}
```

Comment:
:sunglasses:

Reply:
To be honest, I copy / paste an example related to file download in the repo.